tests/common/include/tests/common/tools_rk_stepper.inl

```diff
@@ -0,0 +1,111 @@
+/** Detray library, part of the ACTS project (R&D line)
+ *
+ * (c) 2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#include <gtest/gtest.h>
+
+#include "detray/definitions/units.hpp"
+#include "detray/field/constant_magnetic_field.hpp"
+#include "detray/tools/rk_stepper.hpp"
+#include "detray/tools/track.hpp"
+#include "tests/common/tools/helix_gun.hpp"
+
+/// @note __plugin has to be defined with a preprocessor command
+using namespace detray;
+using namespace __plugin;
+
+constexpr scalar epsilon = 1e-4;
+
+// This tests the base functionality of the Runge-Kutta stepper
+TEST(ALGEBRA_PLUGIN, rk_stepper) {
+
+    // type definitions
+    using vector3 = vector3<scalar>;
+    using point3 = point3<scalar>;
+
+    using mag_field_type = constant_magnetic_field<>;
+    using rk_stepper_type = rk_stepper<mag_field_type, free_track_parameters>;
+
+    // RK stepper configurations
+    constexpr unsigned int theta_steps = 100;
+    constexpr unsigned int phi_steps = 100;
+    constexpr unsigned int max_rk_steps = 1000;
+
+    // Constant magnetic field
+    vector3 B{0, 0, 2 * unit_constants::T};
+    mag_field_type mag_field(B);
+
+    // RK stepper
+    rk_stepper_type rk_stepper(mag_field);
+
+    // Set origin position of tracks
+    const point3 ori{0., 0., 0.};
+    scalar p_mag = 10;
+
+    // Loops of theta values ]0,pi[
+    for (unsigned int itheta = 0; itheta < theta_steps; ++itheta) {
+        scalar theta = 0.1 + itheta * (M_PI - 0.1) / theta_steps;
+        scalar sin_theta = std::sin(theta);
+        scalar cos_theta = std::cos(theta);
+
+        // Loops of phi values [-pi, pi]
+        for (unsigned int iphi = 0; iphi < phi_steps; ++iphi) {
+
+            // The direction
+            scalar phi = -M_PI + iphi * (2 * M_PI) / phi_steps;
+            scalar sin_phi = std::sin(phi);
+            scalar cos_phi = std::cos(phi);
+            const vector3 dir{cos_phi * sin_theta, sin_phi * sin_theta,
+                              cos_theta};
+
+            vector3 mom = p_mag * dir;
+            free_track_parameters traj(ori, 0, mom, -1);
+
+            // helix gun
+            helix_gun helix(traj, B);
+
+            // RK Stepping into forward direction
+            rk_stepper_type::state forward_state(traj);
+            scalar forward_path_limit = 100 * unit_constants::mm;
+            for (unsigned int i_s = 0; i_s < max_rk_steps; i_s++) {
+                rk_stepper.step(forward_state, forward_path_limit);
```

Comment:
I don't know what you meant with initial step size. Anyway I made some changes so take a look

Reply:
I mean the initial step size estimation that is passed to every respective call of the stepper and that then is being evaluated by ```try_rk4``` the first time it is called in a step (so inside ```step(state, h)```). Meaning, the initial step size estimation is the one given to the RK algorithm from the outside, from which it determines if we can safely take that step or if we have to reduce it due to integration errors, so ```h_0``` if you want.  Now, I haven't run an example of this, so I might not understand what's going on and if I misunderstand something, please forget I said anything, but here are my concerns with the current implementation:
The stepper and the navigator are called one after the other by the propagator. The navigator determines which surface we are likely to encounter next with the stepper and calculates the distance. The first initial step size estimation for the stepper will either be that distance or ```1mm``` if the surface is further away than that. Fine, ```1mm``` is our maximal step size here. Now imagine the surface is just a tiny bit further away than ```1mm```: We take the ```1mm``` step, if the B-field is well behaved, re-navigate and then the initial step size estimation will be the updated one from the navigator, so likely very small. Say, with that we reach the surface and the navigator points us to the next surface which is far away again. If we try to take a step now, the initial step size estimation is the step size from before, because we keep the same stepper state, so it is very small. And since we cannot call ```try_rk4``` with an error below the tolerance, we can never increase the step size again (at least not from what I could see and it seems odd for adaptive RK to not go into step size correction if the error is smaller than the tolerance), so we are literally stuck with mini-steps. Same thing happens if we discover a patch of strong magnetic field gradient, where the RK will automatically reduce the step size to cope with the integration error. Even one better, if we manage to overstep, the step size will become negative and in the next step that we try to take, the initial step size estimation that we take from our stepper state will be negative again :O. We might manage to reverse the stepper direction, right?
So, what I meant was, to use the distance to next, that we get from the navigator before every step as initial step size estimation and not use a default value for the first step and afterwards the previous step size. Something like so ```stepping._step_size = es``` before we call ```try_rk4```, which will then determine if we can safely reach our target in a single step. Since we don't want to take gigantic steps in a gap volume, we need a different ```max_step_size``` somewhere. This should then go into ```stepping._step_size = std::min(stepping._step_size, path_limit);```, but ```path_limit``` should not be the same thing as ```es```. Sorry for the wall of text, but maybe it is clearer now what I meant.

Reply:
So you are talking about step size release which increases the step size to the default or whatever you prefer. Yeah we need to do that. The step size release can be called from navigation status function which tells us that track is on surface. I will do that in the next PR for propagator.

For that purpose, navigator or stepper function should take propagator state which contains navigator state and stepper state  but i will do that in next PR..

Reply:
Also In current implementation, the over step doesn't happen because the step size is limited by the "linear" distance to the next surface.

Reply:
> Is it possible that we might manage to reverse the stepper direction?

It's not difficult.

Reply:
What I meant there was that it might happen unintentionally. You mean, you update the step size in some other part of the code through the state objects, like the stepping helper in ACTS does? Otherwise, I must say I am still confused about how we will increase the step size again, even in between surfaces. I will simply wait and see what you have in mind :)

Reply:
> You mean, you update the step size in some other part of the code through the state objects, like the stepping helper in ACTS does?

Exactly

Reply:
Oh, I see, with the previous changes that makes sense now! We should not forget to adapt the ```line_stepper``` to take a path limit instead of the distance to the next surfaces then, too

Reply:
Thanks for clarifying!